### ISSUE TITLE
Bug: some inputs to scipy.spatial.Voronoi can cause Python crash

### ISSUE THREAD

**Issue by mkhorton** [scipy/scipy#8929]
Possible bug in Voronoi causing Python to crash (`pointer being freed was not allocated`) rather than raising an exception. I'm using Voronoi extensively and it is working for 99% of inputs, but some inputs are causing the crash.

I have no idea what's causing this, code below is the most minimal example I could find.

### Reproducing code example:

```
from scipy.spatial import Voronoi
bad_array = [[5.810789664908795, 5.810787503670133, 5.81078813112434],
 [3.8738661942999943, 3.873864895513413, 3.8738657141149844],
 [3.8738612247717636, 3.8738604872941993, 7.747714497477592],
 [3.8738599747717632, 7.747715737294199, 3.8738601774775905],
 [7.747719404771764, 3.873858487294199, 3.873859437477591],
 [7.747719914771763, 7.7477170072942, 7.747717017477592],
 [7.747724884299995, 7.747721415513413, 3.8738682341149846],
 [7.747726134299994, 3.8738661655134132, 7.747722554114985],
 [3.8738667042999944, 7.747723415513414, 7.747723294114985],
 [5.810790273738582, 5.810788867499981, 9.68464361626145],
 [5.810788513738582, 5.810785597499979, 1.9369317162614497],
 [5.8107890237385815, 9.68464411749998, 5.81078929626145],
 [5.810789763738582, 1.93693034749998, 5.81078603626145],
 [9.684648453738582, 5.81078686749998, 5.81078855626145],
 [1.9369303337385817, 5.8107875974999805, 5.81078677626145],
 [5.810790914908796, 1.936932253670133, 9.684642451124342],
 [5.810788414908795, 9.684642753670133, 1.9369338111243402],
 [9.684647844908795, 5.810785503670132, 1.93693307112434],
 [1.9369314849087955, 5.810789503670134, 9.68464319112434],
 [1.9369309749087955, 1.9369309836701327, 5.81078561112434],
 [9.684648354908795, 9.684644023670131, 5.81079065112434],
 [5.810790174908796, 9.684646023670133, 9.68464571112434],
 [5.810789154908796, 1.9369289836701324, 1.93693055112434],
 [9.684649094908796, 1.9369302536701325, 5.81078739112434],
 [1.9369302349087956, 9.684644753670133, 5.810788871124341],
 [1.9369297249087956, 5.810786233670132, 1.9369312911243401],
 [9.684649604908795, 5.810788773670133, 9.68464497112434],
 [3.873864944299995, 7.747720145513414, 1.1394114984397197e-05],
 [3.8738674442999943, 9.645513413246072e-06, 7.747720034114985],
 [7.747724374299994, 3.873862895513413, 1.0654114984453633e-05],
 [8.01429999413305e-06, 3.8738668955134132, 7.747720774114986],
 [7.747725624299995, 7.645513412966516e-06, 3.8738649741149844],
 [7.747718154771762, 7.747713737294198, 5.117477590399999e-06],
 [6.764299994180373e-06, 7.747722145513414, 3.8738664541149848],
 [7.747721164771763, 3.873861757294199, 11.62157133747759],
 [7.747718664771762, 11.621572257294199, 3.8738626974775903],
 [7.747720654771762, 3.2372941989999997e-06, 7.747713757477591],
 [3.873867954299994, 3.8738681655134135, 11.621577614114985],
 [3.8738617347717637, 7.747719007294199, 11.621572077477591],
 [3.8738607147717636, 1.967294199e-06, 3.8738569174775908],
 [3.873865454299995, 11.621578665513415, 3.873868974114985],
 [3.8738594647717632, 3.873857217294199, 2.5974775904000002e-06],
 [11.621578094771763, 7.747715007294198, 3.8738619574775908],
 [3.8738604847717633, 11.6215742572942, 7.747717757477591],
 [1.794771763e-06, 7.7477177372942, 7.747715237477591],
 [1.2847717630000001e-06, 3.873859217294199, 3.8738576574775903],
 [11.621579344771762, 3.873859757294199, 7.747716277477591],
 [11.621584314299994, 3.8738641655134134, 3.8738674941149847],
 [7.747726644299994, 7.747724685513414, 11.621580134114986],
 [7.747725394299994, 11.621579935513413, 7.747725814114985],
 [11.621584824299994, 7.747722685513414, 7.747725074114984],
 [9.684647203738582, 9.68464211749998, 1.9369342362614497],
 [1.9369315837385814, 1.9369323474999804, 9.68464109626145],
 [9.684649703738582, 1.93693161749998, 9.68464287626145],
 [1.9369290837385815, 9.684642847499981, 1.93693245626145],
 [1.9369308437385817, 9.68464611749998, 9.684644356261451],
 [9.68464794373858, 1.9369283474999794, 1.9369309762614497],
 [9.684648963738582, 9.68464538749998, 9.684646136261449],
 [1.9369298237385817, 1.93692907749998, 1.9369291962614497],
 [5.810787904908795, 5.810784233670132, -1.9369237688756602],
 [5.810791424908795, 5.810790773670134, 13.558500031124343],
 [5.810790404908795, -1.9369262663298674, 5.81078487112434],
 [5.810788924908795, 13.558501273670133, 5.810791391124341],
 [13.558507784908794, 5.810786773670132, 5.81078991112434],
 [-1.9369284550912045, 5.8107882336701335, 5.8107863511243405],
 [3.8738656842999943, 6.375513412913344e-06, 8.134114984453632e-06],
 [7.504299994123936e-06, 8.375513413192899e-06, 3.8738631941149846],
 [6.254299994171258e-06, 3.873863625513413, 8.874114984397195e-06],
 [3.8738624747717627, 5.237294199362047e-06, 11.621568817477591],
 [7.747723634299994, 11.621576665513414, 1.3914114984397196e-05],
 [3.8738587247717624, 11.621570987294199, 5.8574775903027826e-06],
 [3.0447717635567944e-06, 3.873862487294199, 11.621569557477592],
 [7.747727384299994, 1.0915513413299245e-05, 11.621576874114984],
 [2.53477176354768e-06, 3.9672941995309185e-06, 7.747711977477591],
 [11.621583064299994, 7.747719415513413, 1.3174114984453633e-05],
 [11.621577584771762, 3.873856487294198, 4.37747759084574e-06],
 [3.477176319823534e-08, 7.747714467294199, 3.337477590302783e-06],
 [7.747718894771763, -3.270580102819374e-08, 1.8574775908457404e-06],
 [5.447717632073497e-07, 11.621572987294199, 3.8738609174775904],
 [8.524299994142165e-06, 7.747725415513414, 11.621578354114984],
 [7.747720424771764, 11.6215755272942, 11.621574597477592],
 [11.621578834771762, 1.2372941988029346e-06, 3.87385869747759],
 [11.621578604771763, 11.6215735272942, 7.747719537477591],
 [11.621585564299993, 8.915513413019689e-06, 7.747721814114984],
 [11.621579854771763, 7.747718277294199, 11.62157385747759],
 [7.274299994189487e-06, 11.621580665513413, 7.7477240341149844],
 [11.621583574299995, 11.621577935513413, 3.8738707541149844],
 [11.621586074299994, 3.873867435513413, 11.621579394114985],
 [3.8738672142999944, 11.621581935513413, 11.621580874114985],
 [5.810791523738581, 1.9369336174999807, 13.55849793626145],
 [5.810787263738582, 9.68464084749998, -1.9369226037385505],
 [5.810787773738581, 13.55849936749998, 1.93693497626145],
 [5.810791013738582, -1.9369249025000201, 9.684640356261449],
 [1.9369320937385819, 5.81079086749998, 13.55849867626145],
 [9.684646693738582, 5.81078359749998, -1.9369233437385505],
 [9.684647713738581, 13.55850063749998, 5.81079181626145],
 [9.684650213738582, 5.810790137499981, 13.55850045626145],
 [1.9369310737385819, -1.9369261725000202, 5.810783516261449],
 [5.810790783738582, 9.68464738749998, 13.558501196261451],
 [1.9369285737385815, 5.81078432749998, -1.9369251237385505],
 [1.9369295937385815, 13.55850136749998, 5.810790036261451],
 [9.684649193738583, -1.93692690250002, 5.81078529626145],
 [5.8107880037385815, 1.9369270774999796, -1.9369258637385505],
 [13.558506633738581, 5.810784867499979, 1.9369334962614497],
 [-1.9369278462614186, 5.81078959749998, 9.68464183626145],
 [5.810789533738582, 13.55850263749998, 9.68464687626145],
 [13.558507143738582, 9.68464338749998, 5.81079107626145],
 [5.810789253738582, -1.9369281725000203, 1.9369284562614497],
 [-1.9369283562614186, 1.9369310774999802, 5.810784256261449],
 [13.558507883738582, 1.9369296174999797, 5.81078781626145],
 [13.558508393738581, 5.81078813749998, 9.684645396261448],
 [-1.9369290962614183, 9.68464484749998, 5.810787516261451],
 [-1.9369296062614183, 5.810786327499981, 1.93692993626145],
 [9.684646594908795, 9.684640753670132, -1.9369212488756602],
 [1.936932734908795, 1.9369342536701337, 13.558497511124342],
 [9.684647104908795, 13.558499273670131, 1.9369363311243402],
 [1.9369322249087955, -1.9369242663298674, 9.684639931124341],
 [1.9369284749087956, 9.684641483670132, -1.9369230288756596],
 [9.684650854908796, 1.9369335236701328, 13.558499291124342],
 [9.684650344908796, -1.9369249963298674, 9.68464171112434],
 [1.9369289849087952, 13.558500003670133, 1.936934551124341],
 [13.558506534908796, 9.684642023670133, 1.93693559112434],
 [-1.936927205091205, 1.9369329836701334, 9.68464067112434],
 [9.684647334908796, 1.9369269836701322, -1.9369245088756595],
 [1.9369319949087955, 9.684648023670134, 13.558500771124342],
 [9.684648584908796, -1.936928266329867, 1.9369298111243411],
 [1.9369307449087956, 13.558503273670134, 9.68464645112434],
 [1.9369292149087955, 1.9369277136701326, -1.9369262888756602],
 [9.684650114908795, 9.684647293670134, 13.558502551124342],
 [1.9369304649087955, -1.9369275363298675, 1.93692803112434],
 [9.684648864908795, 13.558502543670134, 9.68464823112434],
 [13.558507274908795, 1.9369282536701324, 1.9369323311243412],
 [-1.9369279450912045, 9.684646753670133, 9.684643931124342],
 [13.558509034908797, 1.9369315236701325, 9.68464423112434],
 [-1.9369297050912044, 9.684643483670135, 1.936932031124341],
 [-1.9369289650912045, 1.936929713670133, 1.9369287711243401],
 [13.558508294908796, 9.684645293670131, 9.684647491124341]]
Voronoi(bad_array)
```

### Error message:

```
Python(15648,0x7fffa1f4c380) malloc: *** error for object 0x105e1d9e8: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
fish: 'ipython' terminated by signal SIGABRT (Abort)
```

### Scipy/Numpy/Python version information:

```
1.1.0 1.14.3 sys.version_info(major=3, minor=6, micro=5, releaselevel='final', serial=0)
```

**Comment by mkhorton** [comment]
I just tested the above using pyhull, which worked fine, which makes me think this is not a qhull issue.

**Comment by WarrenWeckesser** [comment]
Some print-debugging shows that the call to `qh_new_qhull()` at https://github.com/scipy/scipy/blob/ced1e7b46469dba66e09baf6a0daae047148a368/scipy/spatial/qhull.pyx#L347 is returning the value 3.

The crash occurs in the call to `self.close()` at https://github.com/scipy/scipy/blob/ced1e7b46469dba66e09baf6a0daae047148a368/scipy/spatial/qhull.pyx#L353

The message returned by `self._message.get()` before the call to `self.close()` is

```
qhull precision warning: 
The initial hull is narrow (cosine of min. angle is 1.0000000000000002).
Is the input lower dimensional (e.g., on a plane in 3-d)?  Qhull may
produce a wide facet.  Options 'QbB' (scale to unit box) or 'Qbb' (scale
last coordinate) may remove this warning.  Use 'Pp' to skip this warning.
See 'Limitations' in qh-impre.htm.
QH6227 
qhull precision error: Only 5 facets remain.  Can not merge another
pair.  The input is too degenerate or the convexity constraints are
too strong.

While executing:  | qhull v Qbb Qc Qz
Options selected for Qhull 2015.2.r 2016/01/18:
  run-id 196706885  voronoi  Qbbound-last  Qcoplanar-keep  Qz-infinity-point
  _pre-merge  _zero-centrum  Qinterior-keep  Pgood  _max-width 15
  Error-roundoff 3.1e-14  _one-merge 2.8e-13  Visible-distance 1.9e-13
  U-coplanar-distance 1.9e-13  Width-outside 3.7e-13  _wide-facet 1.1e-12
  _narrow-hull -2.2e-16
Last point added to hull was p113.  Last merge was #5.

At error exit:

Voronoi diagram by the convex hull of 138 points in 4-d:

  Number of Voronoi regions and at-infinity: 6
  Total number of nearly incident points: 132
  Number of Voronoi vertices: 0
  Number of non-simplicial Voronoi vertices: 2

Statistics for:  | qhull v Qbb Qc Qz

  Number of points processed: 6
  Number of hyperplanes created: 12
  Number of facets in hull: 5
  Number of distance tests for qhull: 373
  Number of distance tests for merging: 28
  Number of distance tests for checking: 0
  Number of merged facets: 5
  Maximum distance of merged point above facet: 6.7 (21330046165178.3x)


precision problems (corrected unless 'Q0' or an error)
      7 flipped facets
      1 degenerate hyperplanes recomputed with gaussian elimination
      2 nearly singular or axis-parallel hyperplanes
      1 zero divisors during back substitute
      1 zero divisors during gaussian elimination
```

So it looks like the problem is in the clean up that occurs when the above condition is detected.

That's all I have time for at the moment; I'll let the qhull gurus take it from here.

**Comment by tylerjereddy** [comment]
The bug report is for SciPy `1.1.0` and Python `3.6.5`, both of which are quite out of date now.

I can't reproduce this anymore on ARM Mac with recent Python and SciPy (`main`) versions, so I think I'll close it. If there's a report of a more recent reproduction I can get traction with, feel free to reopen.
